### PR TITLE
Update esm2_native_te recipe, add configurable streaming dataset and non-torchrun support

### PR DESCRIPTION
**Pull request by pstjohn**
Refactors the esm2_native_te recipe to use a configurable load_dataset call, also centralizes the DistributedConfig definition and provides failsafes if the recipe wasn't called with `torchrun`. 

Finally, moves testing to a dedicated `tests/` folder, and adds tests for distributed dataloading to ensure the rank and worldsize parameters are respected for both iterable and streaming datasets.

## Summary by CodeRabbit

- New Features
  - Distributed-aware data loading supporting both streaming and non-streaming datasets with per-rank separation.
  - Endless training iterator for seamless long-running training.
- Improvements
  - Standardized DDP/FSDP2/MFSDP training loops with bf16, gradient clipping, and warmup scheduler.
  - Tokenization/padding streamlined; batching now pads efficiently via the collator.
- Configuration
  - Switched to model_tag; introduced dataset block (tokenizer, micro_batch_size, workers, max_seq_length, dataset source).
  - Added offline/project options for experiment tracking; increased warmup steps.
- Tests
  - Added coverage for distributed dataloaders (streaming and map) and updated training convergence tests.

### CODE REVIEW

**Review by cspades — APPROVED** [review]

File: bionemo-recipes/recipes/esm2_native_te/train_mfsdp.py [hunk 1]
```diff
@@ -65,42 +49,32 @@ def main(args: DictConfig) -> float | None:
...
-    if "facebook" in args.model_name:
-        config = AutoConfig.from_pretrained(args.model_name, dtype=torch.bfloat16)
-        from transformers.models.esm.modeling_esm import EsmForMaskedLM  # noqa: F401
-
-        with (
-            torch.device("meta") if args.fully_shard_kwargs.get("init_model_with_meta_device", True) else nullcontext()
-        ):
-            model = AutoModelForMaskedLM.from_config(config, attn_implementation="flash_attention_2")
-        del model.esm.contact_head
+    config = AutoConfig.from_pretrained(args.model_tag, trust_remote_code=True, dtype=torch.bfloat16)
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/esm2_native_te/train_mfsdp.py:72]: So if they want to load from their own Safetensors file then we need to have NVEsmConfig.from_pretrained here right?

Comment (pstjohn) [reply bionemo-recipes/recipes/esm2_native_te/train_mfsdp.py:72]: yeah exactly, if we don't copy esm_nv.py into their checkpoint directory 🤔 . Also would need NVEsmForMaskedLM in the next line

Comment (jomitchellnv) [reply bionemo-recipes/recipes/esm2_native_te/train_mfsdp.py:72]: Why don't we just let them save an ESM_nv.py file and then they can use the `AutoModel.from_pretrained(<path_to_safetensors_folder>)`

**Comment by pstjohn** [comment]
force merge with recipes CI passing to avoid container tests
